lib/broccoli/broccoli-typescript.js

```diff
@@ -1,170 +1,333 @@
+    if (tsEmitInternal) {
+        var orignalgetCompilerOptions_1 = host.getCompilerOptions;
+        host.getCompilerOptions = function () {
+            var options = clone(orignalgetCompilerOptions_1.call(host));
+            options.stripInternal = false;
+            options.outDir = options.outDir + "/" + exports.INTERNAL_TYPINGS_PATH;
+            return options;
+        };
     }
-  });
+    return originalEmitFiles(resolver, host, targetSourceFile);
+};
+/**
+ * Broccoli plugin that implements incremental Typescript compiler.
+ *
+ * It instantiates a typescript compiler instance that keeps all the state about the project and
+ * can re-emit only the files that actually changed.
+ *
+ * Limitations: only files that map directly to the changed source file via naming conventions are
+ * re-emitted. This primarily affects code that uses `const enum`s, because changing the enum value
+ * requires global emit, which can affect many files.
+ */
+class DiffingTSCompiler extends Plugin {
+    constructor(inputPath, options) {
+        super([inputPath], {});
 
-  Object.keys(this.fileRegistry).forEach(fileName => {
-    if (rootFileNames.indexOf(fileName) === -1) {
-      filesToRemove.push(fileName);
+        this.options = options;
+        this.fileRegistry = Object.create(null);
+        this.firstRun = true;
+        this.previousRunFailed = false;
+        // Whether to generate the @internal typing files (they are only generated when `stripInternal` is
+        // true)
+        this.genInternalTypings = false;
+        if (options.files) {
+            this.rootFilePaths = options.files.splice(0);
+        }
+        else {
+            this.rootFilePaths = [];
+        }
+        if (options.internalTypings) {
+            this.genInternalTypings = true;
+            delete options.internalTypings;
+        }
     }
-  });
 
-  this.tsServiceHost.setRootFileNames(rootFileNames);
+    createServiceHost() {
+        // the conversion is a bit awkward, see https://github.com/Microsoft/TypeScript/issues/5276
+        // in 1.8 use convertCompilerOptionsFromJson
+        this.tsOpts = ts.parseJsonConfigFileContent(this.options, null, null).options;
+        if (this.tsOpts.stripInternal === false) {
+            // @internal are included in the generated .d.ts, do not generate them separately
+            this.genInternalTypings = false;
+        }
+        this.tsOpts.rootDir = '';
+        this.tsOpts.outDir = '';
 
-  filesToEmit.forEach(fileName => {
-    this.emitFile(fileName, this.inputPaths[0], this.outputPath);
-  });
+        this.tsServiceHost = new CustomLanguageServiceHost(
+          this.tsOpts, this.rootFilePaths, this.fileRegistry, this.inputPaths[0]);
+        this.tsService = ts.createLanguageService(this.tsServiceHost, ts.createDocumentRegistry());
+    }
 
-  filesToRemove.forEach(fileName => {
-    this.removeFile(fileName, this.inputPaths[0], this.outputPath);
-  });
-};
+    build() {
```

Comment:
I would use `this.listEntries()` in `build()` method, this is new in broccoli 1.0 and it simplifies things a lot. 
With that method you get times when file was last changed and compare them to previous times, if there is no time in object yet it means it is the initial build. Thats how I did it. Not saying it is the best solution but it is definitely less complex and even more faster. 

Please see https://github.com/jkuri/angular-cli/blob/broccoli-1.0beta/lib/broccoli/broccoli-typescript.js#L32-#L65